File: algorithms/code/leetcode/lc374_guess_number_higher_or_lower/LC374GuessNumberHigherOrLower.java
package leetcode.lc374_guess_number_higher_or_lower;

/**
 * https://leetcode.com/problems/guess-number-higher-or-lower/ #easy
 */
public final class LC374GuessNumberHigherOrLower {
    public int x;

    protected int guess(final int n) {
        if (x < n) {
            return -1;
        }
        if (x > n) {
            return 1;
        }
        return 0;
    }

    public int guessNumber(final int n) {
        int a = 1;
        int z = n;
        while (a != z) {
            int checking = (int) (((long) a + (long) z) / 2);
            switch (guess(checking)) {
            case 1:
                a = checking + 1;
                break;
            case -1:
                z = checking - 1;
                break;
            default:
                return checking;
            }
        }
        return a;
    }
}
package leetcode.lc374_guess_number_higher_or_lower;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public final class LC374GuessNumberHigherOrLowerTests {
    @Test
    public void test2In10() throws Exception {
        LC374GuessNumberHigherOrLower solution = new LC374GuessNumberHigherOrLower();
        solution.x = 2;
        assertEquals(2, solution.guessNumber(10));
    }

    @Test
    public void test8In10() throws Exception {
        LC374GuessNumberHigherOrLower solution = new LC374GuessNumberHigherOrLower();
        solution.x = 8;
        assertEquals(8, solution.guessNumber(10));
    }

    @Test
    public void test65789() throws Exception {
        LC374GuessNumberHigherOrLower solution = new LC374GuessNumberHigherOrLower();
        solution.x = 65789;
        assertEquals(solution.x, solution.guessNumber(Integer.MAX_VALUE));
    }

    @Test
    public void test1() throws Exception {
        LC374GuessNumberHigherOrLower solution = new LC374GuessNumberHigherOrLower();
        solution.x = 1;
        assertEquals(solution.x, solution.guessNumber(1));
    }
}

File: algorithms/code/leetcode/lc374_guess_number_higher_or_lower/lc374_guess_number_higher_or_lower.rs
// https://leetcode.com/problems/guess-number-higher-or-lower/
// #easy

static mut LC374_X: i32 = 0;

pub struct Solution;

impl Solution {
    unsafe fn guess(n: i32) -> i32 {
        if LC374_X < n {
            return -1;
        }
        if LC374_X > n {
            return 1;
        }
        0
    }

    #[allow(clippy::missing_safety_doc)]
    pub unsafe fn guess_number(n: i32) -> i32 {
        let mut a = 1;
        let mut z = n;
        while a != z {
            let checking = ((a as i64 + z as i64) / 2) as i32;
            match Self::guess(checking) {
                1 => {
                    a = checking + 1;
                }
                -1 => {
                    z = checking - 1;
                }
                _ => {
                    return checking;
                }
            }
        }
        a
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_2_in_10() {
        unsafe {
            LC374_X = 2;
            assert_eq!(Solution::guess_number(10), 2);
        }
    }

    #[test]
    fn test_8_in_10() {
        unsafe {
            LC374_X = 8;
            assert_eq!(Solution::guess_number(10), 8);
        }
    }

    #[test]
    fn test_65789() {
        unsafe {
            LC374_X = 65789;
            assert_eq!(Solution::guess_number(i32::MAX), 65789);
        }
    }

    #[test]
    fn test_1() {
        unsafe {
            LC374_X = 1;
            assert_eq!(Solution::guess_number(1), 1);
        }
    }
}
